versioned_docs/version-1.5(WIP)/gwScripts.md

```diff
@@ -0,0 +1,204 @@
+    custodian_script_type_hash: Byte32,
+    deposit_script_type_hash: Byte32,
+    withdrawal_script_type_hash: Byte32,
+    challenge_script_type_hash: Byte32,
+    stake_script_type_hash: Byte32,
+    l2_sudt_validator_script_type_hash: Byte32,
+    burn_lock_hash: Byte32,
+    required_staking_capacity: Uint64,
+    challenge_maturity_blocks: Uint64,
+    finality_blocks: Uint64,
+    reward_burn_rate: byte, // * reward_burn_rate / 100
+    chain_id: Uint64, // chain id
+    allowed_eoa_type_hashes: AllowedTypeHashVec, // list of script code_hash allowed an EOA(external owned account) to use
+    allowed_contract_type_hashes: AllowedTypeHashVec, // list of script code_hash allowed a contract account to use
+  }
+  ````
+
+
+
+The `lock` field of Rollup cell has its own rules comparatively. We assumed in the initial design that everyone who stakes can submit to the Rollup. But during the preliminary phase, we want a more stable setup in which only the block producer can submit to the rollup.
+
+### Stake Lock
+
+A block producer has to provide a stake cell to perform the `RollupSubmitBlock` action. `StakeLockArgs` is the stake lock args. After submitting a layer-2 block, the `args.stake_block_number` will be updated to the latest block number.
+
+  ```js
+  struct StakeLockArgs {
+    owner_lock_hash: Byte32,
+    stake_block_number: Uint64,
+}
+  ````
+
+
+There are two ways to unlock stake lock:
+
+1. Unlocked by the submitter after the `args.stake_block_number` block is finalized.
+2. Unlocked by the challenger during the `RollupRevert` action.
+
+### Deposit Lock
+
+A layer1 user can join the Rollup by creating a deposit cell. Godwoken collects deposit cells from the layer1 blockchain and puts them into the input of the tx which submits layer-2 block.
+
+If the deposit is not processed by Godwoken, the sender can unlock a deposit cell after `cancel_timeout`. 
+
+  ```js
+  table DepositLockArgs {
+    // layer1 lock hash
+    owner_lock_hash: Byte32,
+    layer2_lock: Script,
+    cancel_timeout: Uint64,
+    registry_id: Uint32,
+  }
+  ````
+
+
+### Custodian Lock
+
+Rollup uses the custodian lock to hold the deposited assets. `CustodianLockArgs` is the args of custodian lock, the field `deposit_block_number` represents the block number that the deposit is processed. The `deposit_block_number` also denotes whether the custodian lock is finalized or unfinalized.
+
+```js
+table CustodianLockArgs {
+    deposit_block_hash: Byte32,
+    deposit_block_number: Uint64,
+    // used for revert this cell to deposit request cell
+    // after finalize, this lock is meaningless
+    deposit_lock_args: DepositLockArgs,
+}
+
+struct UnlockCustodianViaRevertWitness {
+    deposit_lock_hash: Byte32,
+}
+```
+
+For unfinalized custodian cells, once the deposit block is reverted, these cells must be also reverted to the deposit cells.
+For finalized custodian cells, since they are finalized, we can free merge or split these cells.
```

Comment:
- Since finalized custodian cells are finalized, we can free merge or split these cells.